Implemented `MultiBindingExpression`

grokys opened the pull request:
## What does the pull request do?

Implements `MultiBindingExpression` and makes `MultiBinding` use it instead of relying on the now-obsolete APIs it was using before.

Shouldn't have much user-facing effect, though does allow us to fix #16084

Also removes a couple of usages of the obsolete `IBinding.Initiate` API from multi-binding tests and adds nullable annotations to that file.

## Fixed issues

Fixes #16084

grokys committed:
Update BindingBase.Instance signature.

- Swap `target` and `targetProperty` order to make it consistent with other similar methods
- Make `targetProperty` nullable as it will need to be null for `MultiBinding`

grokys committed:
IBinding2.Instance needs to accept a null target property.

It will need to be null for `MultiBinding`.

grokys committed:
Attach needs to accept a null target property.

It will need to be null for `MultiBinding`.

grokys committed:
Initial implementation of MultiBindingExpression.

grokys committed:
Fix failing template binding test.

Only publish unset value if we've already published a value.

grokys committed:
Enabled nullability annotations.

grokys committed:
Added passing test for #16084.

grokys committed:
Remove obsolete API usages.

MrJul commented on src/Avalonia.Base/Data/Core/MultiBindingExpression.cs:
```diff
@@ -0,0 +1,131 @@
+                converted = _targetNullValue;
+            if (converted == AvaloniaProperty.UnsetValue)
+                converted = _fallbackValue;
+            PublishValue(converted);
```
This doesn't seem to handle `BindingOperations.DoNothing` anymore: `DoNothing` is directly published to the target.

`MultiValueConverterTests` is supposed to check this but the test passes for the wrong reason, because `DoNothingType` isn't convertible to the target `string` type.

grokys replied:
Good catch! Fixed the test and the issue.

grokys committed:
Merge branch 'master' into feature/multibindingexpression

grokys committed:
Bind to Tag not Text.

Prevents test passing when it shouldn't. See https://github.com/AvaloniaUI/Avalonia/pull/16219#discussion_r1665466968

grokys committed:
Handle DoNothing in MultiBindingExpression.
